# NYC Education Equity

## Importing libraries for data cleaning and analysis

In [1]:
import pandas as pd

## English Language Arts (ELA) Test Results from 2013 to 2023
Source: NYC Open Data

In [2]:
ela_scores = pd.read_csv("../data/raw/English_Language_Arts__ELA__Test_Results_2013-2023_20250828.csv")

In [3]:
ela_scores.head()

Report Category Geographic Subdivision School Name Grade  Year  \
0        Citywide               Citywide         NaN     3  2023   
1        Citywide               Citywide         NaN     4  2023   
2        Citywide               Citywide         NaN     5  2023   
3        Citywide               Citywide         NaN     6  2023   
4        Citywide               Citywide         NaN     7  2023   

  Student Category  Number Tested Mean Scale Score Num Level 1 Pct Level 1  \
0     All Students          49453              447       14225        28.8   
1     All Students          51008              450       10679        20.9   
2     All Students          53235              449       12486        23.5   
3     All Students          51996              447       13237        25.5   
4     All Students          53472              450       12976        24.3   

  Num Level 2 Pct Level 2 Num Level 3 Pct Level 3 Num Level 4 Pct Level 4  \
0       11486        23.2       14154        28.6        9588        19.4   
1       13530        26.5       14788        29.0       12011        23.5   
2       14190        26.7       16717        31.4        9842        18.5   
3       13882        26.7       13709        26.4       11168        21.5   
4       12860        24.0       15838        29.6       11798        22.1   

  Num Level 3 and 4 Pct Level 3 and 4  
0             23742              48.0  
1             26799              52.5  
2             26559              49.9  
3             24877              47.8  
4             27636              51.7

In [4]:
ela_scores.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 626462 entries, 0 to 626461
Data columns (total 18 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Report Category         626462 non-null  object
 1   Geographic Subdivision  626462 non-null  object
 2   School Name             588222 non-null  object
 3   Grade                   626462 non-null  object
 4   Year                    626462 non-null  int64 
 5   Student Category        626462 non-null  object
 6   Number Tested           626462 non-null  int64 
 7   Mean Scale Score        626462 non-null  object
 8   Num Level 1             626462 non-null  object
 9   Pct Level 1             626462 non-null  object
 10  Num Level 2             626462 non-null  object
 11  Pct Level 2             626462 non-null  object
 12  Num Level 3             626462 non-null  object
 13  Pct Level 3             626462 non-null  object
 14  Num Level 4             626462 non-n

All categories except "School Name" are filled with non-null values. Since we are attempting to analyze school-specific data, we will drop the rows containing null values.

In [5]:
ela_scores = ela_scores.dropna()

In [6]:
ela_scores.info()

<class 'pandas.core.frame.DataFrame'>
Index: 588222 entries, 46 to 626461
Data columns (total 18 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Report Category         588222 non-null  object
 1   Geographic Subdivision  588222 non-null  object
 2   School Name             588222 non-null  object
 3   Grade                   588222 non-null  object
 4   Year                    588222 non-null  int64 
 5   Student Category        588222 non-null  object
 6   Number Tested           588222 non-null  int64 
 7   Mean Scale Score        588222 non-null  object
 8   Num Level 1             588222 non-null  object
 9   Pct Level 1             588222 non-null  object
 10  Num Level 2             588222 non-null  object
 11  Pct Level 2             588222 non-null  object
 12  Num Level 3             588222 non-null  object
 13  Pct Level 3             588222 non-null  object
 14  Num Level 4             588222 non-null 

For the ELA Test Scores data, we have 588,222 rows with non-null values. Next, we will drop columns that we won't be using for our analysis, for example, the report category, number and percentage levels.

In [7]:
ela_scores = ela_scores.drop(columns=["Report Category", 
                                      "Geographic Subdivision", 
                                      "Grade",
                                      "Num Level 1", 
                                      "Pct Level 1",
                                      "Num Level 2", 
                                      "Pct Level 2",
                                      "Num Level 3", 
                                      "Pct Level 3",
                                      "Num Level 4", 
                                      "Pct Level 4",
                                      "Num Level 3 and 4", 
                                      "Pct Level 3 and 4",
                                     ])

In [8]:
ela_scores.tail()

School Name  Year Student Category  \
626457  EVERGREEN MIDDLE SCHOOL FOR URBAN EXPLORATION  2014              SWD   
626458  EVERGREEN MIDDLE SCHOOL FOR URBAN EXPLORATION  2013          Not SWD   
626459  EVERGREEN MIDDLE SCHOOL FOR URBAN EXPLORATION  2013              SWD   
626460  EVERGREEN MIDDLE SCHOOL FOR URBAN EXPLORATION  2013          Not SWD   
626461  EVERGREEN MIDDLE SCHOOL FOR URBAN EXPLORATION  2013              SWD   

        Number Tested Mean Scale Score  
626457             45              256  
626458            110              278  
626459             33              256  
626460            110              278  
626461             33              256

In [9]:
ela_scores["Mean Scale Score"].values

array(['s', 's', 's', ..., '256', '278', '256'],
      shape=(588222,), dtype=object)

At this point, we can see **several problems** with the dataset.
1. The "Mean Scale Score" row contains non-number values like 's'. Such rows should be removed from the analysis. 

2. Since the dataset contains ELA test scores from 2013 to 2023, each school repeats up to 11 times. We need to find the mean score for each school across the scores from 2013 and 2023 and aggregate it into a single row.

In [10]:
# Convert column to numeric, invalid entries become NaN
ela_scores["Mean Scale Score"] = pd.to_numeric(ela_scores["Mean Scale Score"], errors="coerce")

ela_scores = ela_scores.dropna()

In [16]:
ela_scores = ela_scores.groupby("School Name", as_index=True).agg({
    "Number Tested": "sum",
    "Mean Scale Score": "mean"
})

In [17]:
ela_scores.head()

Number Tested  \
School Name                                                         
A.C.E. ACADEMY FOR SCHOLARS AT THE GERALDINE FERRA          17057   
ACADEMY FOR COLLEGE PREPARATION AND CAREER EXPLORA          11024   
ACADEMY FOR NEW AMERICANS                                    3849   
ACADEMY FOR PERSONAL LEADERSHIP AND EXCELLENCE              47214   
ACADEMY FOR YOUNG WRITERS                                   13029   

                                                    Mean Scale Score  
School Name                                                           
A.C.E. ACADEMY FOR SCHOLARS AT THE GERALDINE FERRA        459.882911  
ACADEMY FOR COLLEGE PREPARATION AND CAREER EXPLORA        422.003663  
ACADEMY FOR NEW AMERICANS                                 352.265306  
ACADEMY FOR PERSONAL LEADERSHIP AND EXCELLENCE            416.742424  
ACADEMY FOR YOUNG WRITERS                                 436.164062

In [30]:
# Save to a file
ela_scores.to_csv("../data/processed/ela_scores_processed.csv", index=True)


# Math Test Results from 2013 to 2023
Source: NYC Open Data

Let's perform data cleaning similar to the one we did with ELA Test Scores dataset.

In [13]:
math_scores = pd.read_csv("../data/raw/Math_Test_Results_2013-2023_20250902.csv")

In [14]:
math_scores.head()

Report Category Geographic Subdivision School Name Grade  Year  \
0        Citywide               Citywide         NaN     3  2023   
1        Citywide               Citywide         NaN     4  2023   
2        Citywide               Citywide         NaN     5  2023   
3        Citywide               Citywide         NaN     6  2023   
4        Citywide               Citywide         NaN     7  2023   

  Student Category  Number Tested Mean Scale Score Num Level 1 Pct Level 1  \
0     All Students          51866              454        8193        15.8   
1     All Students          53195              453       13460        25.3   
2     All Students          55137              452       15055        27.3   
3     All Students          53519              450       16049        30.0   
4     All Students          54141              454       11723        21.7   

  Num Level 2 Pct Level 2 Num Level 3 Pct Level 3 Num Level 4 Pct Level 4  \
0       15155        29.2       18037        34.8       10481        20.2   
1       11898        22.4       18201        34.2        9636        18.1   
2       12241        22.2       17086        31.0       10755        19.5   
3       13094        24.5       15880        29.7        8496        15.9   
4       14693        27.1       14445        26.7       13280        24.5   

  Num Level 3 and 4 Pct Level 3 and 4  
0             28518              55.0  
1             27837              52.3  
2             27841              50.5  
3             24376              45.5  
4             27725              51.2

In [15]:
math_scores.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 624572 entries, 0 to 624571
Data columns (total 18 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Report Category         624572 non-null  object
 1   Geographic Subdivision  624572 non-null  object
 2   School Name             586383 non-null  object
 3   Grade                   624572 non-null  object
 4   Year                    624572 non-null  int64 
 5   Student Category        624572 non-null  object
 6   Number Tested           624572 non-null  int64 
 7   Mean Scale Score        624572 non-null  object
 8   Num Level 1             624572 non-null  object
 9   Pct Level 1             624572 non-null  object
 10  Num Level 2             624572 non-null  object
 11  Pct Level 2             624572 non-null  object
 12  Num Level 3             624572 non-null  object
 13  Pct Level 3             624572 non-null  object
 14  Num Level 4             624572 non-n

All categories except "School Name" are filled with non-null values. Since we are attempting to analyze school-specific data, we will drop the rows containing null values.

In [19]:
math_scores = math_scores.dropna()

In [20]:
math_scores.info()

<class 'pandas.core.frame.DataFrame'>
Index: 586383 entries, 46 to 624571
Data columns (total 18 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Report Category         586383 non-null  object
 1   Geographic Subdivision  586383 non-null  object
 2   School Name             586383 non-null  object
 3   Grade                   586383 non-null  object
 4   Year                    586383 non-null  int64 
 5   Student Category        586383 non-null  object
 6   Number Tested           586383 non-null  int64 
 7   Mean Scale Score        586383 non-null  object
 8   Num Level 1             586383 non-null  object
 9   Pct Level 1             586383 non-null  object
 10  Num Level 2             586383 non-null  object
 11  Pct Level 2             586383 non-null  object
 12  Num Level 3             586383 non-null  object
 13  Pct Level 3             586383 non-null  object
 14  Num Level 4             586383 non-null 

For the Math Test Scores data, we have 586,383 rows with non-null values. Next, we will drop columns that we won't be using for our analysis, for example, the report category, number and percentage levels.

In [21]:
math_scores = math_scores.drop(columns=["Report Category", 
                                      "Geographic Subdivision", 
                                      "Grade",
                                      "Num Level 1", 
                                      "Pct Level 1",
                                      "Num Level 2", 
                                      "Pct Level 2",
                                      "Num Level 3", 
                                      "Pct Level 3",
                                      "Num Level 4", 
                                      "Pct Level 4",
                                      "Num Level 3 and 4", 
                                      "Pct Level 3 and 4",
                                     ])

In [22]:
math_scores.tail()

School Name  Year Student Category  \
624567  BRONX CHARTER SCHOOL FOR THE ARTS  2015     All Students   
624568  BRONX CHARTER SCHOOL FOR THE ARTS  2016     All Students   
624569  BRONX CHARTER SCHOOL FOR THE ARTS  2017     All Students   
624570  BRONX CHARTER SCHOOL FOR THE ARTS  2018     All Students   
624571  BRONX CHARTER SCHOOL FOR THE ARTS  2019     All Students   

        Number Tested Mean Scale Score  
624567            142              323  
624568            143              320  
624569            142              321  
624570            144              604  
624571            261              597

In [23]:
math_scores["Mean Scale Score"].values

array(['s', 's', 's', ..., '321', '604', '597'],
      shape=(586383,), dtype=object)

At this point, we can see **several problems** with the dataset.
1. The "Mean Scale Score" row contains non-number values like 's'. Such rows should be removed from the analysis. 

2. Since the dataset contains Math test scores from 2013 to 2023, each school repeats up to 11 times. We need to find the mean score for each school across the scores from 2013 and 2023 and aggregate it into a single row.

In [24]:
# Convert column to numeric, invalid entries become NaN
math_scores["Mean Scale Score"] = pd.to_numeric(math_scores["Mean Scale Score"], errors="coerce")

math_scores = math_scores.dropna()

In [25]:
math_scores = math_scores.groupby("School Name", as_index=True).agg({
    "Number Tested": "sum",
    "Mean Scale Score": "mean"
})

In [26]:
math_scores.head()

Number Tested  \
School Name                                                         
A.C.E. ACADEMY FOR SCHOLARS AT THE GERALDINE FERRA          17371   
ACADEMIC LEADERSHIP CHARTER SCHOOL                           3478   
ACADEMY FOR COLLEGE PREPARATION AND CAREER EXPLORA          11019   
ACADEMY FOR NEW AMERICANS                                   13867   
ACADEMY FOR PERSONAL LEADERSHIP AND EXCELLENCE              48644   

                                                    Mean Scale Score  
School Name                                                           
A.C.E. ACADEMY FOR SCHOLARS AT THE GERALDINE FERRA        460.314815  
ACADEMIC LEADERSHIP CHARTER SCHOOL                        426.153846  
ACADEMY FOR COLLEGE PREPARATION AND CAREER EXPLORA        416.498182  
ACADEMY FOR NEW AMERICANS                                 391.558233  
ACADEMY FOR PERSONAL LEADERSHIP AND EXCELLENCE            409.538847

In [31]:
# Save to a file
math_scores.to_csv("../data/processed/math_scores_processed.csv", index=True)
